File: examples/python/secure_communication.py
"""
Secure Communication Example for IMCP Protocol

This module demonstrates secure communication between an IMCP server and client
using public-key cryptography for authentication and encryption.
"""

import asyncio
import os
from nacl.signing import SigningKey  # PyNaCl library for cryptographic operations
from loguru import logger  # Advanced logging library
from imcp.client import IMCPClient
from imcp.server import IMCPServer

async def run_secure_server():
    """
    Initialize and run a secure IMCP server with cryptographic key pair.
    
    This function:
    1. Generates a new Ed25519 signing key pair for the server
    2. Creates an IMCP server instance with the private key
    3. Starts the server and listens for secure connections
    """
    # Generate server's Ed25519 key pair for secure communication
    server_private_key = SigningKey.generate()
    logger.info("Generating server key pair...")
    
    # Initialize IMCP server with secure configuration
    server = IMCPServer(
        host="localhost",  # Server hostname
        port=8766,         # Server port
        private_key=server_private_key.encode()  # Server's private key for secure communication
    )
    logger.info("Starting secure IMCP server...")
    await server.start()

async def run_secure_client():
    """
    Initialize and run a secure IMCP client with cryptographic key pair.
    
    This function demonstrates:
    1. Client key pair generation
    2. Secure connection establishment
    3. Encrypted message exchange
    4. Secure control command handling
    """
    # Generate client's Ed25519 key pair for secure communication
    client_private_key = SigningKey.generate()
    logger.info("Generating client key pair...")
    
    # Initialize IMCP client with secure configuration
    client = IMCPClient(
        "ws://localhost:8766",  # WebSocket server URL
        private_key=client_private_key.encode()  # Client's private key for secure communication
    )
    
    try:
        # Establish secure WebSocket connection to server
        await client.connect()
        logger.info("Secure connection established")

        # Example: Send encrypted sensitive data
        sensitive_data = b"Confidential information"
        await client.send(sensitive_data)
        logger.info("Encrypted message sent")

        # Receive and decrypt response from server
        response = await client.receive()
        logger.info("Encrypted response received")

        # Example: Send secure control command with specific operation
        await client.send_control(
            "SECURE_OPERATION",  # Control command type
            {
                "operation": "encrypt",      # Operation to perform
                "data": "sensitive payload", # Data to process
                "algorithm": "AES-256"       # Encryption algorithm to use
            }
        )
        logger.info("Secure control command processed")

        # Receive secure response for the control command
        secure_response = await client.receive()
        logger.info("Secure response received")

    except Exception as e:
        logger.error(f"Secure client error: {e}")
    finally:
        # Ensure proper cleanup of client resources
        await client.close()

async def main():
    """
    Main function to orchestrate the secure communication example.
    
    This function:
    1. Starts the secure server in the background
    2. Waits for server initialization
    3. Runs the secure client
    4. Properly shuts down the server
    """
    # Start secure server as a background task
    server_task = asyncio.create_task(run_secure_server())
    
    # Allow time for server initialization
    await asyncio.sleep(1)
    
    # Run the secure client
    await run_secure_client()
    
    # Cleanup: Cancel and wait for server task
    server_task.cancel()
    try:
        await server_task
    except asyncio.CancelledError:
        pass

if __name__ == "__main__":
    # Configure logging with rotation and level settings
    logger.add(
        "imcp_secure_example.log",  # Log file name
        rotation="1 MB",            # Rotate log file when it reaches 1MB
        level="INFO"                # Log level
    )
    
    # Run the secure communication example
    asyncio.run(main()) 

File: examples/rust/secure_communication.rs
// Import necessary modules from the IMCP (Inter-Module Communication Protocol) crate
use imcp::client::IMCPClient;
use imcp::protocol::{IMCPError, IMCPHandler, IMCPProtocol, Message};
use imcp::server::IMCPServer;
use log::{error, info};
use std::collections::HashMap;
use std::sync::Arc;
use tokio;

// Define a handler struct that will process incoming messages
// This struct implements the IMCPHandler trait to handle different types of messages
struct SecureHandler;

// Implement the IMCPHandler trait for SecureHandler
// This implementation defines how the handler processes different types of messages
#[async_trait::async_trait]
impl IMCPHandler for SecureHandler {
    // Handle incoming messages based on their type
    // Parameters:
    // - message: The incoming message to process
    // - client_public_key: The public key of the client for verification
    // Returns: Result containing either a response message or an error
    async fn handle_message(
        &self,
        message: Message,
        _client_public_key: &[u8],
    ) -> Result<Message, IMCPError> {
        match message.header.message_type {
            // Handle data messages by creating a response with the same header and body
            imcp::protocol::MessageType::Data => {
                let response = Message {
                    header: message.header,
                    body: message.body,
                    signature: Vec::new(),
                };
                Ok(response)
            }
            // Handle control messages by creating a response with the same header and body
            imcp::protocol::MessageType::Control => {
                let response = Message {
                    header: message.header,
                    body: message.body,
                    signature: Vec::new(),
                };
                Ok(response)
            }
            // Return an error for any unsupported message types
            _ => Err(IMCPError::ProtocolError("Unsupported message type".into())),
        }
    }
}

// Main function that demonstrates secure communication between client and server
#[tokio::main]
async fn main() -> Result<(), IMCPError> {
    // Initialize logging system
    env_logger::init();

    // Create and start the secure server
    // Wrap the handler in an Arc for thread-safe sharing
    let handler = Arc::new(SecureHandler);
    let mut server = IMCPServer::new(handler)?;

    // Spawn the server in a separate task
    let server_handle = tokio::spawn(async move {
        if let Err(e) = server.start("localhost", 8766).await {
            error!("Server error: {}", e);
        }
    });

    // Wait for the server to initialize
    tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;

    // Create and connect the client to the server
    let mut client = IMCPClient::new()?;
    client.connect("ws://localhost:8766").await?;

    // Send sensitive data to the server
    let sensitive_data = b"Confidential information".to_vec();
    client.send(sensitive_data).await?;

    // Receive and log the encrypted response from the server
    let response = client.receive().await?;
    info!("Received encrypted response");

    // Send a secure control command with encryption parameters
    let mut params = HashMap::new();
    params.insert("operation".to_string(), "encrypt".to_string());
    params.insert("data".to_string(), "sensitive payload".to_string());
    params.insert("algorithm".to_string(), "AES-256".to_string());
    client.send_control("SECURE_OPERATION", params).await?;

    // Receive and log the secure response
    let secure_response = client.receive().await?;
    info!("Received secure response");

    // Close the client connection
    client.close().await?;

    // Stop the server
    server_handle.abort();

    Ok(())
}
